### ISSUE TITLE
[FEATURE] Support for custom types in openai tool/function interface

### ISSUE THREAD

**Issue by limoiie** [langchain4j/langchain4j#896]
**Is your feature request related to a problem? Please describe.**

Currently, only Java's built-in types (boolean, byte, short, int, long, float, double, BigDecimal, BigInteger, String, Array, and Array-like) are supported as parameters for tools or functions. Still, custom types may be needed for convenience.

*Example*: Suppose I am developing database tools; each tool works with a specific column type. Without custom types, I would need to repeatedly specify database, table, and column names in each tool's interface. 
```java
class DatabaseToolkit {
    @Tool
    void listByDatetime(
        @P("Database name, it can be ....")
        String dbName,
        @P("Table name, it can be ....")
        String tableName,
        @P("Column name, it can be ....")
        String columnName,
        long start, long end) {
        // ...
    }

    @Tool
    void listByMatch(
        // bad: repeated definition and description
        @P("Database name, it can be ....")
        String dbName,
        @P("Table name, it can be ....")
        String tableName,
        @P("Column name, it can be ....")
        String columnName,
        String pattern) {
        // ...
    }
   
    // more repeated... 
}
```

By organizing these names in a custom type like Column, I can avoid repeating them in every tool or function interface. 

```java
class Column {
        @P("Database name, it can be ....")
        String dbName;
        @P("Table name, it can be ....")
        String tableName;
        @P("Column name, it can be ....")
        String columnName;
        // ...
}

class DatabaseToolkit {
    @Tool
    void listByDatetime(
        Column column,
        long start, long end) {
        // ...
    }

    @Tool
    void listByMatch(
        Column column,
        String pattern) {
        // ...
    }
   
    // more repeated... 
}
```

Additionally, I can manipulate multiple columns at once by using a list of Column instances. Achieving this functionality would be challenging without custom types.

**Describe the solution you'd like**

The current implementation describes the tool or function interface, including all its parameters, in [JsonSchema](https://json-schema.org/understanding-json-schema/), as required by [OpenAI doc](https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools). Python LangChain does so using [pydantic](https://github.com/pydantic/pydantic). By following this convention, we can support custom type to be tool/function parameters in two steps:
- Build a JsonSchema Generator to convert correctly annotated custom type into JsonSchema;
- Build a JsonSchema Validator to validate and coerce the argument generated by LanguageModel into the custom type.

To do so, one can extend [ToolSpecifications](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java) by generating JsonSchema for custom types. To bind JsonSchema attributes to the custom type and its fields, new annotations can be introduced, or existing annotations like [Jackson-Annotations](https://github.com/FasterXML/jackson-annotations/wiki/Jackson-Annotations) can be adopted directly. In [DefaultToolExecutor](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java), one should add code to validate the arguments according to the JsonSchema or the corresponding annotations, and then coerce them to corresponding types, including custom ones.

Several points to consider:
-  Composition: When a custom type is derived, we should recursively compose the schemas of its superclasses using the `allOf` attribute in JsonSchema.
-  Polymorphism: If a custom type is abstract or a superclass that can accept instances of its subclasses, we should use the `oneOf` attribute in JsonSchema.

*Benefit*: This solution is lightweight and requires minimal changes to the current code. It also avoids adding additional dependencies to the langchain4j core library.

*Shortcoming*: It may reinvent the wheel.

*NOTE*: Supporting custom type parameters will break the token counting process of [OpenAiTokenier](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiTokenizer.java) as it assumes the parameters can only be Java built-in types when estimating tokens in tool specification. Once we support custom type, we should update OpenAiTokenizer's code logic.

**Describe alternatives you've considered**

The proposed solution might be reinventing the wheel since there are many open-source JsonSchema generators in Java. Here are several alternatives:
- The [victools/jsonschema-generator](https://github.com/victools/jsonschema-generator) tool supports generating and validating JsonSchema with respect to annotations from Jackson, Jakarta, javax, and Swagger.
- The [networknt/json-schema-validator](https://github.com/networknt/json-schema-validator) tool supports the [JSON Schema Core Draft v4, v6, v7, v2019-09 and v2020-12](http://json-schema.org/latest/json-schema-core.html) specification for JSON schema validation with respect to Jackson.

*Benefit*: One big benefit of adopting these existing approaches is that once validated, you can directly deserialize the JSON string back to beans since they respect serialization frameworks like Jackson.

*Shortcoming*: However, using existing JsonSchema generators may introduce more dependencies to the core library, and their features could be overly broad for the tool/function calling situation.

**Additional context**

My concerns include:
- Does langchain4j really need custom-type parameters?
- If so, is there a current process to support custom-type parameters?
- If not, what is the preferred approach: implementing a lightweight JsonSchema directly into langchain4j (similar to the initial solution) or using an existing JsonSchema Generator (like the other options)?

**Comment by langchain4j** [comment]
Thanks a lot!

A few thoughts:
- While it is true that supporting custom POJOs as `@Tool` method parameters can simplify many things for devs, it is usually recommended to flatten the schema for function calling as it is way easier for the LLM to adhere to flat schemas than deep nested ones. This might not be true in the long run, so it is still a good feature to include, put probably not the highest prio.
- Regarding reinventing the wheel, I totally agree. Current implementation is buggy and lacks a lot of features. It was created with a very simple use case in mind and now has grown into a big messy pile of code. I guess it is a good idea to look into existing libraries to delegate this functionality. Regarding extra dependency concern: we could make it as a separate pluggable dependency via SPI. So by default, only existing primitive logic is executed (without custom POJOs). If user defines custom POJO as a tool parameter, an exception is thrown that says "... please add langchain4j-tools-schema-generator dependency ..." or something like that.

WDYT?

Partially addressed in https://github.com/langchain4j/langchain4j/pull/708
Related to https://github.com/langchain4j/langchain4j/issues/859

**Comment by sbeugen** [comment]
Hi @langchain4j ,

I would be interested if List parameters would also be covered by the issues you linked. At the moment I get the error 
`array schema missing items ...`

The workaround is to tell the AI to pass the values as a comma separated list in a single string.

**Comment by limoiie** [comment]
> Thanks a lot!
> 
> A few thoughts:
> 
> * While it is true that supporting custom POJOs as `@Tool` method parameters can simplify many things for devs, it is usually recommended to flatten the schema for function calling as it is way easier for the LLM to adhere to flat schemas than deep nested ones. This might not be true in the long run, so it is still a good feature to include, put probably not the highest prio.
> * Regarding reinventing the wheel, I totally agree. Current implementation is buggy and lacks a lot of features. It was created with a very simple use case in mind and now has grown into a big messy pile of code. I guess it is a good idea to look into existing libraries to delegate this functionality. Regarding extra dependency concern: we could make it as a separate pluggable dependency via SPI. So by default, only existing primitive logic is executed (without custom POJOs). If user defines custom POJO as a tool parameter, an exception is thrown that says "... please add langchain4j-tools-schema-generator dependency ..." or something like that.
> 
> WDYT?
> 
> Partially addressed in #708 Related to #859

Thank you for your excellent work! Sorry for the delay. I find the SPI solution excellent. I've submitted a PR that delegates `ToolSpecification` generation and `ToolExecutionRequest` deserialization to `ToolJsonSchemas` with `JsonSchemaService`. The `VictoolsJsonSchemaService` implements `JsonSchemaService` using [victools/jsonschema-generator](https://github.com/victools/jsonschema-generator) for a more complete JSON schema capability. Apart from SPI, one can also provide an implementation for `JsonSchemaService` via `AiServices#jsonSchemaService`.
